# [Return to Workflow](#eod)

# Phase 2 Project - DSC 060721 <a name="top"></a>

Author: Ben McCarty

Date: 7/12/21

Contact: bmccarty505@gmail.com

# 🧾 Table of Contents: <a name="toc"></a>

- ☑ [To-Do's](#todo)
- 📃 [Rubric](#Rubric)
- 🏡 [Title](#title)


- 📂 [Setup](#setup)
- 🔮 [Exploring Data](#explore)
- 🛠 [Feature Engineering](#eng)
- 🔗 [Correlations](#correlations)


- 🚿 [Pipeline_1](#pipeline_1)
- 🚿 [Pipeline_2](#pipeline_2)


- 🥒 [Pickling](#pickle)

# **☑** **To-Do's** **☑** <a name='todo'></a>

- [ ] Fix [guiding questions](#td-questions)
- [x] Modularize functions - incl. H2 read/inspect notes
- [x] Create 'report_df" function
- [ ] Add links to each section addressing the rubric
- [ ] Use pipeline for inferential and predictive steps
- [ ] Use GridSearchCV for hyperparameter tuning
- [ ] H2: add VIF to pipeline?
- [ ] Feature Engine: OHE
- [ ] Lasso Reg in pipeline

# 📃 Rubric <a name="rubric"></a>

---
**Business Understanding:** [Business Understanding](#bizund)

- [ ] Notebook clearly explains the project's value for helping a specific stakeholder solve a real-world problem.


- [ ] Introduction explains the real-world problem the project aims to solve


- [ ] Introduction identifies stakeholders who could use the project and how they would use it


- [ ] Conclusion summarizes implications of the project for the real-world problem and stakeholders

---
**Data Understanding:**
- [ ] Notebook clearly describes the source and properties of the data to show how useful the data are for solving the problem of interest.


- [ ] Describe the data sources and explain why the data are suitable for the project


- [ ] Present the size of the dataset and descriptive statistics for all features used in the analysis


- [ ] Justify the inclusion of features based on their properties and relevance for the project


- [ ] Identify any limitations of the data that have implications for the project

---
**Data Preparation:** 
- [ ] Notebook shows how you prepare your data and explains why by including…


- [ ] Instructions or code needed to get and prepare the raw data for analysis


- [ ] Code comments and text to explain what your data preparation code does


- [ ] Valid justifications for why the steps you took are appropriate for the problem you are solving

---
**Modeling:**
- [ ] Notebook demonstrates an iterative approach to model building.


- [ ] Runs and interprets a simple, baseline model for comparison


- [ ] Introduces new models that improve on prior models and interprets their results


- [ ] Explicitly justifies model changes based on the results of prior models and the problem context


- [ ] Explicitly describes any improvements found from running new models

---
**Regression Results:**
- [ ] Notebook clearly presents two features with strong relationships to sale prices


- [ ] Explicitly identify the features and interpret their coefficients in plain language


- [ ] Explain the implications of results for the real-world problem the project aims to address


- [ ] Identify specific actions that stakeholders should take to leverage these results

---
**Visualization:**
- [ ] Notebook includes three relevant and polished visualizations of findings that…


- [ ] Help the project stakeholder understand the value or success of the project


- [ ] Have text and marks to aid reader interpretation, such as graph and axis titles, axis ticks and labels, or legend (varies by visualization type)


- [ ] Use color, size, and/or location to appropriately facilitate comparisons


- [ ] Are not cluttered, dense, or illegible

---
**Code Quality:**
- [ ] Code in notebook and related files meets professional standards (e.g. PEP 8)


- [ ] Code is easy to read, using comments, spacing, variable names, and function docstrings


- [ ] All code runs and no code or comments are included that are not needed for the project


- [ ] Code minimizes repetition, using loops, functions, and classes


- [ ] Code adapted from others is properly cited with author names and location of the cited material
---

# 🏡 **(Re)Selling Seattle: Determining House Sell Prices** 🏡 <a name="title"></a>

<a name='bizund'></a>

**<p style="text-align: center;">If you are considering selling your house, one of the highest priorities is getting the best price for it.</p>**

You want to make sure that the hedges are trimmed; that the basement isn't leaking; and that new coat of paint is covering up all of the "art" covering the walls from your kids.

*But what else can do you?* Would that extra bathroom under the stairs be a worthwhile addition? What about that addition you always talked about building?

---
**<p style="text-align: center;">When you are exploring the option of selling, you may ask such questions as:</p>**

>* How do the different aspects of your house impact the price?
>
>
>* What would be a predicted sell price for your house as-is?
>
>
>* What can you do to improve the house's value?

---
You may have some ideas already (and if you don't, watch a few episodes of "This Old House" and you will have *plenty* of inspiration).

But how do you *really* know what features are best? How can you be sure that your intuition or expectations are based on fact and not assumptions? **Turn to the data for answers!**

---
**<p style="text-align: center;">Revisiting King County</p>**

My [prior exploration](https://github.com/BenJMcCarty/Phase_2_Project_Final) of the King County House Sales dataset yielded some recommendations, but I wanted to see if I can improve them. I am revisiting the data with new techniques and approaches to see how much I can improve the results.

---

# BEN - FIX THIS! <a name="td-questions"></a>

**WHAT QUESTIONS ARE IMPORTANT FROM THE BUSINESS PERSPECTIVE? WHAT QUESTIONS DID I USE TO GUIDE MY EXPLORATION?** 

My project used data from house sales in King County, WA. to determine some of the top features determining the sell price of a home.

* **First, I explored the data using the Pandas package to review the data itself.** I reviewed the different features included in my dataset and their respective values; the descriptive statistics for the numerical data; and the overall size and shape of the data (how many rows and columns).

* **Then, I used Seaborn to plot the data and a fitted linear regression model for each feature against the sell price.** These visualizations helped differentiate which features to treat as *categorical* variables (e.g. features that would be one of a select number of options) versus *continuous* variables (which would have less of a rigid range of values).

* **In order to give more depth to my data, I engineered four new features based on the original features and data.** I determined the age of each house at sale; whether or not a home was renovated; how many years since a renovation; and whether or not a house had a basement.

* **I wanted to make sure all of the features and their data were relevant to apply to my future model.** I performed correlational comparisons to determine which features were too closely related (indicating multicollinearity), which would affect my modeling process later on.

* **To test my assumptions that certain features would have a statistically significant impact on the price, I performed a few statistical tests.** I performed an ANOVA test on our "condition" feature (comparing each ranking of condition against the others) and I ran a two-sample t-test on the "waterfront" and "was_renovated" features as well.

* **LReg for Inferences**

# **Setting Up the Tools and Data** <a name='setup'></a>

## Imports

In [19]:
# Data Handling
import pandas as pd
import numpy as np
from scipy import stats

# Visualizations
import matplotlib.pyplot as plt
import seaborn as sns

# Modeling - Statsmodels
import statsmodels.formula.api as smf
import statsmodels.api as sms

# Modeling - SKLearn
from sklearn import metrics
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error
from sklearn.preprocessing import PolynomialFeatures, StandardScaler, OneHotEncoder, FunctionTransformer
from sklearn.model_selection import train_test_split, cross_validate, GridSearchCV
from sklearn.feature_selection import RFECV
from sklearn.impute import SimpleImputer, KNNImputer
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn import set_config

# Settings
%matplotlib inline
plt.style.use('seaborn-talk')
pd.set_option('display.max_columns', None)
pd.set_option('display.float_format', lambda x: f'{x:,.2f}')
set_config(display='diagram')

## Importing Personal Functions

I created several functions specific to this project and saved them in an external file. The file is available from the main directory.

In [20]:
%reload_ext autoreload
%autoreload 2

In [21]:
from p2pr_functions.functions import *

## Reading Data

In [5]:
# old_data = 'https://raw.githubusercontent.com/BenJMcCarty/Phase_2_Project_Final/main/kc_house_data_train.csv'

# new_data = 'https://raw.githubusercontent.com/BenJMcCarty/king_county_housing/main/data/kc_house_data.csv'

local_data = 'data/kc_house_data.csv'

In [6]:
data = pd.read_csv(local_data, index_col=0)

# **Exploring Fresh Data**

## Basic Overviews

The goal for this section is to get a broad idea of the data before I start any cleaning or feature engineering.

In [7]:
## Copying dataframe to new dataframe (in case I need the original later)
df = data

In [8]:
df.head()

date      price  bedrooms  bathrooms  sqft_living  sqft_lot  \
id                                                                              
7129300520  10/13/2014 221,900.00         3       1.00         1180      5650   
6414100192   12/9/2014 538,000.00         3       2.25         2570      7242   
5631500400   2/25/2015 180,000.00         2       1.00          770     10000   
2487200875   12/9/2014 604,000.00         4       3.00         1960      5000   
1954400510   2/18/2015 510,000.00         3       2.00         1680      8080   

            floors  waterfront  view  condition  grade  sqft_above  \
id                                                                   
7129300520    1.00         nan  0.00          3      7        1180   
6414100192    2.00        0.00  0.00          3      7        2170   
5631500400    1.00        0.00  0.00          3      6         770   
2487200875    1.00        0.00  0.00          5      7        1050   
1954400510    1.00        0.00  0.00          3      8        1680   

           sqft_basement  yr_built  yr_renovated  zipcode   lat    long  \
id                                                                        
7129300520           0.0      1955          0.00    98178 47.51 -122.26   
6414100192         400.0      1951      1,991.00    98125 47.72 -122.32   
5631500400           0.0      1933           nan    98028 47.74 -122.23   
2487200875         910.0      1965          0.00    98136 47.52 -122.39   
1954400510           0.0      1987          0.00    98074 47.62 -122.05   

            sqft_living15  sqft_lot15  
id                                     
7129300520           1340        5650  
6414100192           1690        7639  
5631500400           2720        8062  
2487200875           1360        5000  
1954400510           1800        7503

In [9]:
## Review summary of dataframe details
report_df(df)

(21597, 20)


datatypes  num_unique  null_sum  null_pct     count       mean  \
date             object         372         0      0.00       nan        nan   
price           float64        3622         0      0.00 21,597.00 540,296.57   
bedrooms          int64          12         0      0.00 21,597.00       3.37   
bathrooms       float64          29         0      0.00 21,597.00       2.12   
sqft_living       int64        1034         0      0.00 21,597.00   2,080.32   
sqft_lot          int64        9776         0      0.00 21,597.00  15,099.41   
floors          float64           6         0      0.00 21,597.00       1.49   
waterfront      float64           2      2376      0.11 19,221.00       0.01   
view            float64           5        63      0.00 21,534.00       0.23   
condition         int64           5         0      0.00 21,597.00       3.41   
grade             int64          11         0      0.00 21,597.00       7.66   
sqft_above        int64         942         0      0.00 21,597.00   1,788.60   
sqft_basement    object         304         0      0.00       nan        nan   
yr_built          int64         116         0      0.00 21,597.00   1,971.00   
yr_renovated    float64          70      3842      0.18 17,755.00      83.64   
zipcode           int64          70         0      0.00 21,597.00  98,077.95   
lat             float64        5033         0      0.00 21,597.00      47.56   
long            float64         751         0      0.00 21,597.00    -122.21   
sqft_living15     int64         777         0      0.00 21,597.00   1,986.62   
sqft_lot15        int64        8682         0      0.00 21,597.00  12,758.28   

                     std       min        25%        50%        75%  \
date                 nan       nan        nan        nan        nan   
price         367,368.14 78,000.00 322,000.00 450,000.00 645,000.00   
bedrooms            0.93      1.00       3.00       3.00       4.00   
bathrooms           0.77      0.50       1.75       2.25       2.50   
sqft_living       918.11    370.00   1,430.00   1,910.00   2,550.00   
sqft_lot       41,412.64    520.00   5,040.00   7,618.00  10,685.00   
floors              0.54      1.00       1.00       1.50       2.00   
waterfront          0.09      0.00       0.00       0.00       0.00   
view                0.77      0.00       0.00       0.00       0.00   
condition           0.65      1.00       3.00       3.00       4.00   
grade               1.17      3.00       7.00       7.00       8.00   
sqft_above        827.76    370.00   1,190.00   1,560.00   2,210.00   
sqft_basement        nan       nan        nan        nan        nan   
yr_built           29.38  1,900.00   1,951.00   1,975.00   1,997.00   
yr_renovated      399.95      0.00       0.00       0.00       0.00   
zipcode            53.51 98,001.00  98,033.00  98,065.00  98,118.00   
lat                 0.14     47.16      47.47      47.57      47.68   
long                0.14   -122.52    -122.33    -122.23    -122.12   
sqft_living15     685.23    399.00   1,490.00   1,840.00   2,360.00   
sqft_lot15     27,274.44    651.00   5,100.00   7,620.00  10,083.00   

                       max  
date                   nan  
price         7,700,000.00  
bedrooms             33.00  
bathrooms             8.00  
sqft_living      13,540.00  
sqft_lot      1,651,359.00  
floors                3.50  
waterfront            1.00  
view                  4.00  
condition             5.00  
grade                13.00  
sqft_above        9,410.00  
sqft_basement          nan  
yr_built          2,015.00  
yr_renovated      2,015.00  
zipcode          98,199.00  
lat                  47.78  
long               -121.31  
sqft_living15     6,210.00  
sqft_lot15      871,200.00

## Data Cleaning and Processing

In [ ]:
## Creating a pairplot instead of the individual plots below
# sns.pairplot(data=df,corner=True)
# plt.show()

---
**DF Columns to Convert**

* 'date' to datetime; continuous variable


Potential Changes:

>* 'waterfront' to category; nominal/binary variable
>* 'view' to category; ordinal
>* 'condition' to category; ordinal variable
>* 'grade' to category; ordinal variable
>* 'zipcode' to category; nominal variable


---
**Null Values to Process**

Columns with null values:
* 'yr_renovated'
* 'waterfront
* 'view'

Options:
>* Fill with mean/median/mode
>* Fill with zeroes
>
>
>* **PIPELINES!!!**
>    * SimpleImputer
>    * IterativeImputer *(Experimental)*
>    * GridSearchCV for hyperparameter tuning

---

In [14]:
## Converting 'date' column to datetime

df['date'] = pd.to_datetime(df['date'])
df['date']

id
7129300520   2014-10-13
6414100192   2014-12-09
5631500400   2015-02-25
2487200875   2014-12-09
1954400510   2015-02-18
                ...    
263000018    2014-05-21
6600060120   2015-02-23
1523300141   2014-06-23
291310100    2015-01-16
1523300157   2014-10-15
Name: date, Length: 21597, dtype: datetime64[ns]

In [15]:
## Converting 'sqft_basement' to numeric and filling any null values with zero

df['sqft_basement'] = pd.to_numeric(df['sqft_basement'], errors='coerce')
df['sqft_basement']

id
7129300520     0.00
6414100192   400.00
5631500400     0.00
2487200875   910.00
1954400510     0.00
              ...  
263000018      0.00
6600060120     0.00
1523300141     0.00
291310100      0.00
1523300157     0.00
Name: sqft_basement, Length: 21597, dtype: float64

In [16]:
df_na_col = []
for col in df:
    if df[col].isna().sum() > 0:
        df_na_col.append(col)

df_na_col

[]

In [17]:
for i in df_na_col:
    df[i].fillna(df[i].mode()[0], inplace=True)

In [18]:
report_df(df)

(21597, 20)


datatypes  num_unique  null_sum  null_pct     count  \
date           datetime64[ns]         372         0      0.00       nan   
price                 float64        3622         0      0.00 21,597.00   
bedrooms                int64          12         0      0.00 21,597.00   
bathrooms             float64          29         0      0.00 21,597.00   
sqft_living             int64        1034         0      0.00 21,597.00   
sqft_lot                int64        9776         0      0.00 21,597.00   
floors                float64           6         0      0.00 21,597.00   
waterfront            float64           2         0      0.00 21,597.00   
view                  float64           5         0      0.00 21,597.00   
condition               int64           5         0      0.00 21,597.00   
grade                   int64          11         0      0.00 21,597.00   
sqft_above              int64         942         0      0.00 21,597.00   
sqft_basement         float64         303         0      0.00 21,597.00   
yr_built                int64         116         0      0.00 21,597.00   
yr_renovated          float64          70         0      0.00 21,597.00   
zipcode                 int64          70         0      0.00 21,597.00   
lat                   float64        5033         0      0.00 21,597.00   
long                  float64         751         0      0.00 21,597.00   
sqft_living15           int64         777         0      0.00 21,597.00   
sqft_lot15              int64        8682         0      0.00 21,597.00   

                    mean        std       min        25%        50%  \
date                 nan        nan       nan        nan        nan   
price         540,296.57 367,368.14 78,000.00 322,000.00 450,000.00   
bedrooms            3.37       0.93      1.00       3.00       3.00   
bathrooms           2.12       0.77      0.50       1.75       2.25   
sqft_living     2,080.32     918.11    370.00   1,430.00   1,910.00   
sqft_lot       15,099.41  41,412.64    520.00   5,040.00   7,618.00   
floors              1.49       0.54      1.00       1.00       1.50   
waterfront          0.01       0.08      0.00       0.00       0.00   
view                0.23       0.76      0.00       0.00       0.00   
condition           3.41       0.65      1.00       3.00       3.00   
grade               7.66       1.17      3.00       7.00       7.00   
sqft_above      1,788.60     827.76    370.00   1,190.00   1,560.00   
sqft_basement     285.72     439.82      0.00       0.00       0.00   
yr_built        1,971.00      29.38  1,900.00   1,951.00   1,975.00   
yr_renovated       68.76     364.04      0.00       0.00       0.00   
zipcode        98,077.95      53.51 98,001.00  98,033.00  98,065.00   
lat                47.56       0.14     47.16      47.47      47.57   
long             -122.21       0.14   -122.52    -122.33    -122.23   
sqft_living15   1,986.62     685.23    399.00   1,490.00   1,840.00   
sqft_lot15     12,758.28  27,274.44    651.00   5,100.00   7,620.00   

                     75%          max  
date                 nan          nan  
price         645,000.00 7,700,000.00  
bedrooms            4.00        33.00  
bathrooms           2.50         8.00  
sqft_living     2,550.00    13,540.00  
sqft_lot       10,685.00 1,651,359.00  
floors              2.00         3.50  
waterfront          0.00         1.00  
view                0.00         4.00  
condition           4.00         5.00  
grade               8.00        13.00  
sqft_above      2,210.00     9,410.00  
sqft_basement     550.00     4,820.00  
yr_built        1,997.00     2,015.00  
yr_renovated        0.00     2,015.00  
zipcode        98,118.00    98,199.00  
lat                47.68        47.78  
long             -122.12      -121.31  
sqft_living15   2,360.00     6,210.00  
sqft_lot15     10,083.00   871,200.00

### Overview Summary

The dataset contains 20 columns of data, most of which seem useful for evaluations and modeling.

I manually fixed null values at this stage - I will start the pipeline later on.

After processing, I do not have any null values and all of the datatypes are correct (I can convert some columns into the "category" datatype for analysis later, if needed).

## Exploring Features

Now I will be comparing each individual feature against "price" to determine if a feature is continuous or categorical.

I will check the regressions and histograms for each feature; if a regression plot shows a flatter regression line or if the histogram is not a normal distribution, I will treat that feature as a categorical feature.

As part of the function to create the plots, I include my function to ignore any outliers for the purpose of the visualizations. I will *not* change the data itself, though.


| Feature Names<a name='features'></a> | Feature Names  |
| ------------ | ------------- |
| [Bedrooms](#bedrooms)     | [Sqft_above](#above)    |
| [Bathrooms](#bathrooms)    | [Sqft_basement](#base) |
| [Sqft_living](#living)  | [Yr_built](#built)      |
| [Sqft_lot](#lot)     | [Yr_renovated](#reno)  |
| [Floors](#floors)       | [Zipcode](#zip)       |
| [Waterfront](#wf)   | [Lat](#lat)           |
| [View](#view)         | [Long](#long)          |
| [Condition](#cond)    | [Sqft_living15](#living15)    |
| [Grade](#grade)        | [Sqft_lot15](#lot15)    |

**Bedrooms** <a name="bedrooms"></a> [🔝](#features)

In [ ]:
show_cleaned_vis(data=df,x='bedrooms', categorical=True, kde = False)

***
**Observations**
>* There is a slight linear regression between the number of bedrooms and the sell price, indicated by the positive slope of the regression line.
>
>
>* The histogram slightly resembles a normal distribution, although the first and last bins are lower than I would expect for a normal distribution.

***
**TO-DO**
>* Use `'bedrooms'` as a categorical variable.
***

**Bathrooms** <a name="bathrooms"></a> [🔝](#features)

In [ ]:
show_cleaned_vis(df,"bathrooms", categorical=True, kde=False)

***
**Observations**
>* There is a clear linear relationship between bathrooms and price.
>
>
>* The histogram/normal distribution isn't the strongest, particularly for the larger number of bathrooms.

***
**TO-DO**
>
>* Similar to bedrooms, the values for number of bathrooms are discrete and fall into specific values. This indicates that the feature is categorical.
*** 

**Sqft_living** <a name="living"></a> [🔝](#features)

In [ ]:
show_cleaned_vis(df,"sqft_living")

***
**Observations**

>* There is a clear linear relationship, although the distribution is slightly skewed to the right, indicating it's not a perfect normal distribution.
>
>
>* As the values are spread across the whole range, this feature is clearly a continuous variable.

***

**To-Do**

>* Use as continuous variable
***


**Sqft_lot** <a name="lot"></a>

In [ ]:
show_cleaned_vis(df,"sqft_lot")

***
**Observations**

>* Weak linear relationship with price; distribution skewed right.
***

**To-Do**

>* Treat as continuous variable as the values are spread out across the whole range.

***

**Floors** <a name="floors"></a> [🔝](#features)

In [ ]:
show_cleaned_vis(df,"floors", kde=False)

***
**Observations**

>* Linear relationship, but clearly categorical as the values fall into specific categories ranging from 1-3.5
>
>* Most homes had one or two floors.

***
**To-Do**

>* Treat as categorical variable.
 
***

**Waterfront** <a name="wf"></a> [🔝](#features)

In [ ]:
feature_vis(df,"waterfront", categorical=True, kde=False)

***
**Observations**

>* Graphs are misleading - clearly binary variable with a slight linear relationship.
>
>
>* Most properties are non-waterfront.
>
>
>* Clear linear relationship, but not normally distributed

***
**To-Do**

>* Treat as categorical - only two options (0 or 1).
 
***

**View** <a name="view"></a> [🔝](#features)

In [ ]:
feature_vis(df,"view", categorical=True, kde=False)

***
**Observations**

>* Slight linear relationship - higher "view" rating, higher price
>
>
>* Most homes have '0' view
>
>
>* A few extreme outliers in pricing (for values 2, 3, and 4s)

***
**To-Do**

>* Treat as categorical
 
***

**Condition** <a name="cond"></a> [🔝](#features)

In [ ]:
show_cleaned_vis(df,"condition", categorical=True, kde= False)

***
**Observations**

>* Largest number of homes sold were in condition 3, followed by condition 4.
>
>
>* Very few homes sold with conditions of 1s or 2s.
>
>
>* Price outliers in 4.0 area, some slight outliers in 2 and 3
>
>
>* No linear relationship, not normal distribution.

***
**TO-DO**

>* Treat as categorical
 
***

**Grade** <a name="grade"></a> [🔝](#features)

In [ ]:
show_cleaned_vis(df,"grade", kde=False)

**Observations**

>* Clear linear trend - as grade increases, so does price
>
>
>* Grades range from 6-9
>
>
>* Price outliers for grades 7 and 8

---
**To-Do**

>* Treat as categorical variable


**Sqft_above** <a name="above"></a> [🔝](#features)

In [ ]:
show_cleaned_vis(df,"sqft_above")

---

**Observations**

>* Outliers impacting accuracy of linear regression, but still seems relatively linear.
>
>
>* Distribution skewed right, but otherwise normal
---
**TO-DO**

>* Treat as continuous
---

**Sqft_basement** <a name="base"></a>[🔝](#features)

In [ ]:
show_cleaned_vis(df,"sqft_basement")

---
**Observations**

>* There is a very large number of 0 ft$^2$ basements, indicating that these homes do not have a basement
>
>
>* There are scattered outliers in pricing
>
>
>* The regression is poor due to the large number of 0 values

---
**To-Do**

>* Treat as a continuous variable
>
>
>* Can create new categorical, binary variable to indicate whether or not a house has a basement to measure the impact on price.
---

**Yr_built** <a name="built"></a> [🔝](#features)

In [ ]:
show_cleaned_vis(df,"yr_built")

---
**Observations**

>* No linear trend from this regression plot
>
>* Seems like may houses built between 1940 - 1970, then major boom in early 2000s.
>
>* Some significant outliers in price, but not enough to  impact the regression
---

**To-Do**

>* Compare year built to year sold - how old was the house at sale?
>
>
>* Potentially discretize for analysis.
---

**Yr_renovated** <a name="reno"></a> [🔝](#features)

In [ ]:
feature_vis(df,'yr_renovated', kde=False)

---
**Observations**

>* Number of houses not renovated significantly outweighs the number renovated
>
>
>* Graphs indicate that an engineered feature for whether or not the house was renovated would be worthwhile.

---
**To-Do**

>* Create new categorical feature - renovated or not
---

**Zipcode** <a name="zip"></a> [🔝](#features)

In [ ]:
show_cleaned_vis(df,"zipcode", categorical=True, kde=False)

---
**Observations**

>* No linear relationship; not normally distributed
---
**TO-DO**

>* Treat as categorical variable
>
>* Perform OneHotEncoding to create new features for modeling.
---

**Lat** <a name="lat"></a> [🔝](#features)

In [ ]:
show_cleaned_vis(df,"lat")

---
**Observations**

>* Very weak linear relationship and not normally distributed.
>
>
>* Outliers just slightly past 47.6 degrees latitude

---
**To-Do**

>* Treat as categorical for modeling.
---

**Long** <a name="long"></a> [🔝](#features)

In [ ]:
show_cleaned_vis(df,"long")

---
**Observations**

>* No linear relationship and not normally distributed.
>
>
>* Outliers around -122.32ish and -122.25-ish degrees

---
**To-Do**

>* Treat as continuous if used for modeling.
---

**Sqft_living15** <a name="living15"></a> [🔝](#features)

In [ ]:
show_cleaned_vis(df,"sqft_living15")

---
**Observations**

>* Most properties sold between 1500/2000 ft$^2$ living space
>
>
>* Price follows linear trend and is normally distributed, although skewed right.
>
>* Outliers between 2500-3500 ft$^2$ 
---
**TO-DO**

>* Treat as continuous variable
---

**Sqft_lot15** <a name="lot15"></a> [🔝](#features)

In [ ]:
show_cleaned_vis(df,"sqft_lot15")

---
**Observations**

>* No clear linear trend
>* Significant outlier at approx. 9000 ft$^2$

---
**TO-DO**
>* Treat as continuous variable
---

# **Feature Engineering**

## `'yrs_old_sold'`

I create this feature to differentiate between houses that were built recently versus older houses.

In order to determine this feature, I need to determine the year the house was sold first.

### Determine `'year_sold'`

In [ ]:
## Pull the year from the "date" column
df['year_sold'] = pd.DatetimeIndex(df['date']).year

## Review the values to ensure data integrity
df['year_sold'].value_counts()

### Calculate `'y_old_sold'`

In [ ]:
## Calculating the age of the house at the time of sale
df['y_old_sold'] = df['year_sold'] - df['yr_built']

## Minimum age is -1 due to a house being sold before it was finished being built
display(df['y_old_sold'].value_counts().sort_index(), df['y_old_sold'].describe())

>**Results:** Half of the homes sold fall between the ages of 18 and 63 years old, with the average age of 43 years. Our min and max ages indicate a house was sold before it was built, while another house was 115 years old.

## `'was_renovated'`

What impact would a renovation have on the price?

In [ ]:
## Using the year that the home was renovated to deterine whether or not the home was renovated
reno_y_n = np.where(df['yr_renovated']>0, 1, 0 )
df = df.assign(was_renovated = reno_y_n)
df.head(5)

In [ ]:
## Checking values
df['was_renovated'].value_counts()

>**Results:** Most of the houses were not renovated at the time of sale. This feature may not have a significant impact on determining the sale price due to the small number of renovated houses.

## `"yrs_since_reno"`

If a house was renovated, how long ago was the renovation? Would more newly-renovated houses increase price?

In [ ]:
## Ensuring there are no null values in the new feature and replacing any with zeroes

df['yrs_since_reno'] = np.where((df['was_renovated']==1),
                                (df['year_sold'] - df['yr_renovated']), 0)

display(df['yrs_since_reno'].describe(),df['yrs_since_reno'].value_counts(ascending=False))

In [ ]:
## Checking for which properties show a renovation year post-sale

df[df['year_sold'] < df['yr_renovated']]

In [ ]:
## Reviewing stats for those houses that were renovated

yrs_and_reno = df[df['yrs_since_reno'] > 0]
yrs_and_reno['yrs_since_reno'].describe()

>**Results:** As expected, most of the houses were not renovated (indicated by the number of years being zero). 
>
>
>Of those renovated, half of the homes had between 9 and 28 years between the years sold and renovated, with an average ages of roughly 20 years old.
>
>
>Interestingly, this dataset also shows *six records for which the renovation is shown to be post-sale.* While this may be an error, I am leaving it in the dataset since I cannot confidently rule it out. 
>
>
>If this feature does not show statistical significance, I may drops these values and reevaluate the significance.

## "`has_bsmnt`"

I noticed that there were fewer houses with a value for "sqft_basement" during my data exploration. I am curious if the presence or absence of a basement would have any impact.

In [ ]:
## Determining whether or not a house has a basement based on the square footage
df['has_bsmnt'] = np.where(df['sqft_basement'] > 0, 1, 0)

# Reviewing the results
display(df['has_bsmnt'].describe(), df['has_bsmnt'].value_counts())

# **Correlations**

## Determining Correlations with Price

In [ ]:
## Determining top five features most strongly correlated with price
## (Considering absolute values for magnitude, not direction)

df_corr = np.abs(df.drop(['price', 'lat','long', 'yr_renovated'], axis=1).corrwith(df['price']).sort_values(ascending=False))
display(df_corr[0:5])

## Determining Multicollinearity

In [ ]:
## Get the correlation matrix for the data (without the target)
corr = np.abs(df.drop('price',axis=1).corr())
corr.round(2)

In [ ]:
corr_val(df.drop(['price','yr_renovated'],axis=1), figsize=(20,20));

In [ ]:
## Correlation results ignoring duplicate values 
## Except for "was_renovated", which shows a perfect relationship to itself

df_corr_results = np.abs(df.drop(['price','yr_renovated'],axis=1).corr().unstack().sort_values(ascending=False).drop_duplicates())

## Show strongest correlations
df_corr_results[1:11]

**Results:** From the correlation matrix, I see that we have seven pairs of features that show a correlation over .75.

I will drop 'sqft_living', 'was_renovated', and 'has_basement' due to their high correlations.

In [ ]:
# Dropping columns to address multicollinearity (over .75)
df_dropped = df.drop(['was_renovated', 'sqft_living','has_bsmnt'], axis=1)
display(df_dropped)

In [ ]:
# Rerunning model
corr_val(df_dropped.drop('price',axis=1), figsize=(20,20));

In [ ]:
# Correlation results ignoring (most) duplicate values
df_corr_results = df_dropped.drop('price',axis=1).corr().unstack().sort_values(ascending=False).drop_duplicates()

## Show strongest correlations
print(df_corr_results[1:11],'\n\n',df_corr_results[-11:-1])

# 🌟 **Creating (Another) Pipeline for Modeling** 🌟 <a name="pipeline_2"><a name='eod'></a> [🔝](#top)

**For original code:** [Click here](#pipeline_1)

## Ex: Creating imputers, encode

In [ ]:
## Instantiate an imputer to add zeroes

imp_zero = SimpleImputer(strategy='constant', fill_value=0)

In [ ]:
## Instantiate an imputer to fill values with the mode value

imp_mode = SimpleImputer(strategy='most_frequent')

In [ ]:
## Instantiate a OneHotEncoder for categorical variables

ohe = OneHotEncoder(drop='first', sparse=False)

## Creating ColumnTransformer

In [ ]:
## Setting up the individual transformers and then the ColumnTransformer

## Adding zeroes for missing values in sqft_basement
imp_zero_cols = ['sqft_basement']
zero_transformer = Pipeline(steps=[
    ('imp_zero',SimpleImputer(strategy='constant', fill_value=0))])


## Adding the mode value to view, waterfront, and yr_renovated missing values 
imp_mode_cols = ['view','waterfront','yr_renovated']
mode_transformer = Pipeline(steps=[
    ('imp_mode',SimpleImputer(strategy='most_frequent'))])


# ## Removed this code due to errors encountered b/t train, test categoricals
# ohe_cols = ['condition','grade', 'zipcode']
# categorical_transformer = OneHotEncoder(drop='first', sparse=False)


## Encoding categoricals - handling errors to prevent issues w/ test set
ohe_cols = ['condition','grade', 'zipcode']
categorical_transformer = OneHotEncoder(handle_unknown='ignore', sparse=False)


## Instantiating the ColumnTransformer to handle the missing values for ea.col.
preprocessor = ColumnTransformer(
    transformers=[
        ('zero', zero_transformer, imp_zero_cols),
        ('mode', mode_transformer, imp_mode_cols),
        ('cats', categorical_transformer, ohe_cols)])

preprocessor

In [ ]:
## Fit and transform the data via the ColumnTransformer
X_train_tf = preprocessor.fit_transform(X_train)

In [ ]:
## Getting feature names from OHE
ohe_cat_names = preprocessor.named_transformers_['cats'].get_feature_names(ohe_cols)
ohe_cat_names

In [ ]:
## Generating list for column index
final_cols = [*imp_zero_cols, *imp_mode_cols, *ohe_cat_names]

In [ ]:
## Instantiating a DF storing the resulting transformed training data
X_train_tf_df = pd.DataFrame(X_train_tf, columns=final_cols, index=X_train.index)
X_train_tf_df

In [ ]:
## Transforming the test set
X_test_tf = preprocessor.transform(X_test)
X_test_tf

In [ ]:
## Creating DF for test data
X_test_tf_df = pd.DataFrame(X_test_tf, columns=final_cols, index=X_test.index)
X_test_tf_df

In [ ]:
## Running a Statsmodels OLS model for verification of R2
sm_reg = sms.OLS(y_train, X_train_tf_df).fit()
sm_reg.summary()

In [ ]:
## Creating baseline parameters
sm_reg.params

## Creating full pipeline

In [ ]:
## Creating full pipeline with ColumnTransformer and LinearRegression
reg_pipe = Pipeline(steps=[('preprocessor', preprocessor),
                      ('regressor', LinearRegression())])

In [ ]:
## Fittng pipeline to the training data
reg_pipe.fit(X_train, y_train)

In [ ]:
## Running the pipeline on the training data
y_hat_train = reg_pipe.predict(X_train)
y_hat_train

In [ ]:
## Running the pipeline on the testing data
y_hat_test = reg_pipe.predict(X_test)
y_hat_test

In [ ]:
reg_pipe['regressor'].intercept_

In [ ]:
pipe_coefs = pd.Series(reg_pipe['regressor'].coef_, index=final_cols)
pipe_coefs['intercept'] = reg_pipe['regressor'].intercept_
pipe_coefs.head(20)

In [ ]:
eval_perf_train(reg_pipe, X_train, y_train)

In [ ]:
eval_perf_test(reg_pipe, X_test, y_test)

In [ ]:
# ## Evaluating performance of training model

# train_mae = metrics.mean_absolute_error(y_train, y_hat_train)
# train_mse = metrics.mean_squared_error(y_train, y_hat_train)
# train_rmse = np.sqrt(metrics.mean_squared_error(y_train, y_hat_train))
# train_r = metrics.r2_score(y_train, y_hat_train)

# print('Evaluating performance of training data:\n')
# print(f'Mean Absolute Error: {train_mae:,.2f}')
# print(f'Mean Squared Error:  {train_mse:,.2f}\n')
# print(f'Root Mean Squared Error: {train_rmse:,.2f}')
# print(f'R-Square value for training data: {round(train_r,2)}')

In [ ]:
# ## Evaluating performance of testing model

# test_mae = metrics.mean_absolute_error(y_test, y_hat_test)
# test_mse = metrics.mean_squared_error(y_test, y_hat_test)
# test_rmse = np.sqrt(metrics.mean_squared_error(y_test, y_hat_test))
# test_r = metrics.r2_score(y_test, y_hat_test)

# print('Evaluating performance of testing data:\n')
# print(f'Mean Absolute Error: {test_mae:,.2f}')
# print(f'Mean Squared Error:  {test_mse:,.2f}\n')
# print(f'Root Mean Squared Error: {test_rmse:,.2f}')
# print(f'R-Square value for testing data: {round(test_r,2)}')

## Creating Example GridSearchCV

In [ ]:
# Creating first GridSearchCV to determine best parameters for LinReg

param_grid = {'fit_intercept': [True, False]}
ex_grid = GridSearchCV(LinearRegression(), param_grid, verbose=True,
                   scoring='r2')
ex_grid.fit(X_train_tf_df, y_train)
display(ex_grid.best_params_)
display(ex_grid.best_score_)
display(ex_grid.best_estimator_)
display(ex_grid.get_params())

In [ ]:
## Results in a DF format
# pd.DataFrame(reg.cv_results_).T

In [ ]:
## Performing GSCV on test data
y_hat_test = ex_grid.predict(X_test_tf_df)
y_hat_test

In [ ]:
eval_perf_test(ex_grid, X_test, y_test)

In [ ]:
# ## Evaluating model peformance

# test_mae = metrics.mean_absolute_error(y_test, y_hat_test)
# test_mse = metrics.mean_squared_error(y_test, y_hat_test)
# test_rmse = np.sqrt(metrics.mean_squared_error(y_test, y_hat_test))
# test_r = metrics.r2_score(y_test, y_hat_test)

# print('Evaluating performance of testing data.')
# print()
# print(f'Mean Absolute Error: {test_mae:,.2f}')
# print(f'Mean Squared Error:  {test_mse:,.2f}')
# print()
# print(f'Root Mean Squared Error: {test_rmse:,.2f}')
# print(f'R-Square value for testing data is {round(test_r,2)}.')

In [ ]:
## Running gridsearch on pipeline

param_grid = {'preprocessor__mode__imp_mode__strategy': ['mean','median',
                                                        'most_frequent']}
gs_imp_rmse = GridSearchCV(reg_pipe, param_grid, verbose=True,
                   scoring='neg_root_mean_squared_error')
gs_imp_rmse.fit(X_train, y_train)
display(gs_imp_rmse.best_score_, reg.best_params_)

In [ ]:
eval_perf_train(gs_imp_rmse, X_train, y_train)

In [ ]:
eval_perf_test(gs_imp_rmse, X_test, y_test)

In [ ]:
## Running gridsearch on pipeline

param_grid = {'preprocessor__mode__imp_mode__strategy': ['mean','median',
                                                        'most_frequent']}
gs_imp_r2 = GridSearchCV(reg_pipe, param_grid, verbose=True,
                   scoring='r2')
gs_imp_r2.fit(X_train, y_train)
display(gs_imp_r2.best_score_)
display(gs_imp_r2.best_params_)

In [ ]:
eval_perf_train(gs_imp_r2, X_train, y_train)

In [ ]:
eval_perf_test(gs_imp_r2, X_test, y_test)

In [ ]:
pd.DataFrame(gs_imp_r2.cv_results_)

# **Modeling with Statsmodels**

## Baseline Model with Raw Data

Creating a baseline model using the raw dataset.

In [ ]:
## Reviewing feature names and their datatypes
df.info()

In [ ]:
## Establishing the features for inferential modeling

categorical_features = ['was_renovated','has_bsmnt', 'waterfront']

continuous_features = ['yrs_since_reno', 'bedrooms', 'bathrooms',
                       'condition','grade', 'floors']

model_baseline = create_model(df,continuous_features,categorical_features);

---
**Interpretation:**

>* R^2: ~.6, less than target of .75
>
>* Residual plots show heteroscedasticity
>
>* Q-Q Plot shows non-normal residuals
>
>* Changes: remove outliers and retest
---

## Model without Outliers in 'Price'

What would the model look like if we ignored the outliers?

### Removing Outliers from 'Price' Data

In [ ]:
## Remove outliers from price 
idx_outs = find_outliers_z(df['price'])
df_clean = df[~idx_outs].copy()
df_clean

In [ ]:
## Remove outliers from bedrooms
idx_outs = find_outliers_z(['bedrooms'])
df_clean = df[~idx_outs].copy()

df_clean

In [ ]:
df_clean.describe()

In [ ]:
## Redefining updated features and re-running model

categorical_features = ['was_renovated','has_bsmnt', 'waterfront']

continuous_features = ['yrs_since_reno', 'bedrooms', 'bathrooms',
                       'condition','grade', 'floors']

model_clean = create_model(df_clean,cont=continuous_features,cat=categorical_features);

Interpretation:

* R^2: ~.59, less than target of .75 (performed slightly worse)

* Residual plots show somewhat increased homoscedasticity

* Q-Q Plot shows more normal residuals (vs. earlier plot)

* Changes: add zipcode 

## Model (Clean) - w/ Zip Codes

In [ ]:
## Adding zip codes as a categorical variable

categorical_features = ['was_renovated','has_bsmnt', 'waterfront', 'zipcode']

continuous_features = ['yrs_since_reno', 'bedrooms', 'bathrooms',
                       'condition','grade', 'floors']

model_w_zip = create_model(df_clean,continuous_features, categorical_features)

Interpretation:

>* Model is significant with an R-squared value over the threshold of .75 (result: .77)
>
>* Residual plots show somewhat homoscedasticity
>
>* Q-Q Plot shows more normal residuals (vs. earlier plot)
>
>* Changes: remove features with p-values higher than .05 (statistically insignificant values)

In [ ]:
## Re-running model after removing 'has_bsmnt' and 'floors'" due to high p-value
categorical_features = ['was_renovated','waterfront', 'zipcode']

continuous_features = ['yrs_since_reno', 'bedrooms', 'bathrooms',
                       'condition','grade']

model_clean = create_model(data=df_clean, cont=continuous_features, cat=categorical_features)

In [ ]:
type(model)

## Recommendations based on Inferential Modeling

If a homeowner is interested in renovating their house, I would recommend the following improvements based on the high coefficients from my model.

* Add bathrooms
* Also consider adding bedrooms
* Use high-quality materials in renovations

While it is outside of the homeowner's ability to change, the zip code for a house's location also has a large impact on the price. A prospective buyer should keep the zip code of the area in mind to help determine the prices of the homes there.

In [ ]:
## Visualizing the impact of number of bathrooms on price
sns.regplot(data=df_clean, x="bathrooms", y='price')
plt.suptitle('Impact of Number of Bathrooms on Price')
plt.xlabel('Number of Bathrooms')
plt.ylabel('Price ($)');

In [ ]:
## Visualizing impact of number of bedrooms on price
sns.regplot(data=df_clean, x="bedrooms", y='price')
plt.suptitle('Impact of Number of Bedrooms on Price')
plt.xlabel('Number of Bedrooms')
plt.ylabel('Price ($)');

In [ ]:
## Visualizing impact of the grade of construction materials on price
sns.regplot(data=df_clean, x="grade", y='price')
plt.suptitle('Impact of Grade of Materials on Price')
plt.xlabel('Grade of Materials')
plt.ylabel('Price ($)');

In [ ]:
## Representing the diversity of prices based on zip codes

fg = sns.catplot(data=df_clean, x="zipcode", y='price', aspect=2.75, height=5)
fg.ax.set_xticklabels(fg.ax.get_xticklabels(), rotation=45, ha='right')
plt.suptitle('Impact of Zip Code on Price')
plt.xlabel('Zip Code')
plt.ylabel('Price ($)');

# **(Old) Modeling with SKLearn**

Now that I have a working model, I want to go ahead and test it to see how well it works. I will split the data in a 75/25 split into two groups; train a model on the 75%; and then compare it against the remaining 25%. 

**My goal is to have a Root Mean Squared Error (RMSE) less than 200,000.**

## ID Features, Target

First, I want to establish the features I want to use in my testing and my target variable (in this case, price).

In [ ]:
# ## Save the features that we used previously - I will use these features 
# ## to fit my model.

# X = df[['was_renovated','waterfront', 'zipcode',
#             'yrs_since_reno', 'bedrooms', 'bathrooms', 'condition','grade']].copy()

In [ ]:
# ## Creating the y values by setting them equal to the 'price' values from the dataframe

# y = df['price'].copy()

In [ ]:
# ## Verifying the two groups are of equal length

# print(X.shape[0] == y.shape[0])

## Create Train/Test Split

Now I will split the data into the train/test groups. Then, I will run the first linear regression on the "train" data, then another regression on the "test" data.

In [ ]:
# ## Establishing the train and test data before doing anything else

# X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25,
#                                                     random_state=5)

In [ ]:
# ## Using OneHotEncoder to create columns for the categorical variables.

# ## Create list of categorical variables
# cat_feats = ['condition','grade', 'zipcode']

# ## Create the OHE without "drop='first" as it would throw an error in this case
# ohe = OneHotEncoder(handle_unknown = 'ignore', sparse=False)

# ## Using OHE on our categorical variables for training (NOT testing)
# train_ohe = ohe.fit_transform(X_train[cat_feats])
# test_ohe = ohe.transform(X_test[cat_feats])

In [ ]:
# ## Getting feature names from our list of categories
# feat_col_name = ohe.get_feature_names(cat_feats)

In [ ]:
# ## Creating dataframes from the results of our fit&transform and transform

# train_ohe_df = pd.DataFrame(train_ohe, columns=feat_col_name, index=X_train.index)

# test_ohe_df = pd.DataFrame(test_ohe, columns=feat_col_name, index=X_test.index)

In [ ]:
# ## Reviewing one of the resulting dataframes
# test_ohe_df

In [ ]:
# ## Reviewing the training dataset
# X_train

In [ ]:
# ## Concatenating the two training dataframes after OHE
# X_train_ohe = pd.concat([X_train.drop(cat_feats, axis=1), train_ohe_df], axis=1)

In [ ]:
# ## Concatenating the two test dataframes after OHE
# X_test_ohe = pd.concat([X_test.drop(cat_feats, axis=1), test_ohe_df], axis=1)
# X_test_ohe

## Fitting the Model

Now I will create the linear regression model.

In [ ]:
# ## Instantiating the model and fitting it
# lr = LinearRegression()

# lr.fit(X_train_ohe, y_train)

In [ ]:
# ## Checking the R^2 for train and test

# train_r = lr.score(X_train_ohe, y_train)
# print(f'R-Square value for training data is {round(train_r,3)}.')

# test_r = lr.score(X_test_ohe, y_test)
# print(f'R-Square value for test data is {round(test_r,3)}.')

# ## Getting model coefficients
# train_coef = pd.Series(lr.coef_, index=X_train_ohe.columns)
# train_coef['intercept'] = lr.intercept_

# ## Displaying resulting features and coefficients
# train_coef.sort_values(ascending=False)#.plot(kind='barh');

In [ ]:
# ## Determining predictions for training data

# y_train_pred = lr.predict(X_train_ohe)

# ## Determining predictions for test data

# y_test_pred = lr.predict(X_test_ohe)

## Evaluate Performance

Let's see how it performed against the test dataset!

In [ ]:
# eval_perf_train(lr,X_train, y_train)

In [ ]:
# ## Calculating scores for test model

# test_mae = metrics.mean_absolute_error(y_test, y_test_pred)
# test_mse = metrics.mean_squared_error(y_test, y_test_pred)
# test_rmse = np.sqrt(metrics.mean_squared_error(y_test, y_test_pred))

# print(f'Mean Absolute Error: {test_mae:,.2f}')
# print(f'Mean Squared Error: {test_mse:,.2f}')
# print()
# print(f'Root Mean Squared Error: {test_rmse:,.2f}')
# test_r = lr.score(X_test_ohe, y_test)
# print(f'R-Square value for test data is {round(test_r,2)}.')

## Comparing Results

In [ ]:
# ## Calculating difference between RMSE(Training) and RMSE(Test)

# print('Our RMSE values for each set:')
# print(f'Train RMSE: {train_rmse:,.2f}')
# print(f'Test RMSE: {test_rmse:,.2f}')

# diff = train_rmse - test_rmse

# print()
# print(f'The difference in RMSE between our training and test sets is {diff:,.2f}.')

## ✨ Verifying Assumptions ✨

In [ ]:
# ## Determining residuals
# residuals = (y_test - y_test_pred)
# residuals

In [ ]:
# ## Plotting to test for normality
# sns.histplot(data=residuals);

In [ ]:
# ## Checking the homoscedasticity of the new model
# sns.residplot(x=y_test, y=residuals, lowess=True, color="g");

### Interpretation of Model Results

The model does a great job of explaining the variance between our selected features, with an r-squared value of .79/1.

The RMSE for both the training and test sets are below our target of 200,000! Additionally, the difference between the two is within an acceptable range.

The assumption tests confirm the assumptions that our residuals would be normally distributed and homoscedastic.

## Recursive Feature Elimination

I have a good model created above, but I'd like to see if changing the features would have an impact for me.

**I will use the Recursive Feature Elimination method with Cross Validation to determine another selection of features.**

In [ ]:
# ## Creating the settings for the selector
# selector = RFECV(estimator=lr, step=2, cv=3, scoring='neg_mean_squared_error', n_jobs=-1, verbose=1)

In [ ]:
# ## Running it on our training data
# selector.fit(X_train_ohe, y_train)

In [ ]:
# ## Dividing the data into two separate variables for inspection
# selected_columns = X_train_ohe.columns[selector.get_support()]
# removed_columns = X_train_ohe.columns[~selector.get_support()]

In [ ]:
# ## Inspecting the columns with the greatest impact according to the RSECV method
# list(selected_columns.values)

In [ ]:
# ## Testing the new list of columns

# ## Instantiating a new regression model
# lr_rfe = LinearRegression()

# ## Fitting the model to the new data
# lr_rfe.fit(X_train_ohe[selected_columns], y_train)

# # Using the model to create new predictions
# y_rfe = lr_rfe.predict(X_train_ohe[selected_columns])

# # Determining RMSE for new list of columns and TRAIN data
# trainRFE_rmse = np.sqrt(metrics.mean_squared_error(y_train, y_rfe))

# print(f'Training Root Mean Squared Error: {trainRFE_rmse:,.2f}')

# # Determining RMSE for new list of columns and TEST data
# y_pred_rfe = lr_rfe.predict(X_test_ohe[selected_columns])

# testRFE_rmse = np.sqrt(metrics.mean_squared_error(y_test, y_pred_rfe))

# print(f'Testing Root Mean Squared Error: {testRFE_rmse:,.2f}')

# diff2 = trainRFE_rmse - testRFE_rmse
# print()
# print(f'The difference in RMSE between our training and test sets is {diff2:,.2f}.')

In [ ]:
# test_rsq = lr_rfe.score(X_test_ohe[selected_columns], y_test)
# print(f'R-Square value for testing data after RFECV is {round(test_rsq,2)}.')

### Interpretation of Results of RFEVC

---
**Comparing RMSEs from original model to RFECV:**

* **Pre-RFECV model results:**
   * The R^2 value for both training and testing data was .79 (significant as it is over the .70 threshold).
   * *Our RMSE values for each set:*
        * Train RMSE: 117,679.79
        * Test RMSE: 122,263.56
    * The difference in RMSE between our training and test sets is **-4,583.77**.


 * **Post-RFECV model results:**
     * The R^2 value for the test data is .48 (.31 lower than other model, and far below threshold).
     * *Our RMSE values for each set:*
         * Training Root Mean Squared Error: 182,158.41
         * Testing Root Mean Squared Error: 191,360.88
     * The difference in RMSE between our training and test sets is **-9,202.47**.

---

**Final Analysis of RFE method:** The RFE results produced a large decrease in model perfomance as determined by the lower R^2 and lower RMSE values. I will continue using the first model for future predictions.

---

## Model Testing - Original Data 

The model produced above was based on data from which we removed any outliers. I am re-running the tests to determine if there is any difference in the results. 

**My goal is to have a Root Mean Squared Error (RMSE) less than 200,000.**

In [ ]:
# ## Load, recast dtypes

# df= pd.read_csv(local_data,parse_dates = ['date'], index_col=0)
# df.info()

In [ ]:
# df['sqft_basement'] = pd.to_numeric(df['sqft_basement'], errors='coerce')

In [ ]:
# df.info()

## ID Features, Target

In [ ]:
# cols = ['was_renovated','waterfront', 'zipcode',
#             'yrs_since_reno', 'bedrooms', 'bathrooms', 'condition','grade']

# [col for col in df.columns if col not in cols]

In [ ]:
# drop_cols = ['date', 'lat','long', 'price']

In [ ]:
# ## Saving the features that I used for inferential modeling - I will use these features 
# ## to fit my model.

# X = df.drop(columns = drop_cols).copy()

In [ ]:
# X.head()

In [ ]:
# ## Creating the y values by setting them equal to the 'price' values from the dataframe

# y = df['price'].copy()

In [ ]:
# ## Verifying the two groups are of equal length

# print(X.shape[0] == y.shape[0])

## Create Train/Test Split

Now I will split the data into the train/test groups. Then, I will run the first linear regression on the "train" data, then another regression on the "test" data.

In [ ]:
# ## Establishing the train and test data before doing anything else

# X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25,
#                                                     random_state=5)

In [ ]:
# X_train.isna().sum()

# 🌟 **Creating the Pipeline** 🌟 <a name='pipeline_1'></a> [🔝](#top)

In [ ]:
## Instantiate an imputer to add zeroes

imp_zero = SimpleImputer(strategy='constant', fill_value=0)

In [ ]:
## Instantiate an imputer to fill values with the mode value

imp_mode = SimpleImputer(strategy='most_frequent')

In [ ]:
## Instantiate a OneHotEncoder for categorical variables

ohe = OneHotEncoder(drop='first', sparse=False)

In [ ]:
## Setting up the individual transformers and then the ColumnTransformer

## Adding zeroes for missing values in sqft_basement
imp_zero_cols = ['sqft_basement']
zero_transformer = Pipeline(steps=[
    ('imp_zero',SimpleImputer(strategy='constant', fill_value=0))])


## Adding the mode value to view, waterfront, and yr_renovated missing values 
imp_mode_cols = ['view','waterfront','yr_renovated']
mode_transformer = Pipeline(steps=[
    ('imp_mode',SimpleImputer(strategy='most_frequent'))])


# ## Removed this code due to errors encountered b/t train, test categoricals
# ohe_cols = ['condition','grade', 'zipcode']
# categorical_transformer = OneHotEncoder(drop='first', sparse=False)


## Encoding categoricals - handling errors to prevent issues w/ test set
ohe_cols = ['condition','grade', 'zipcode']
categorical_transformer = OneHotEncoder(handle_unknown='ignore', sparse=False)


## Instantiating the ColumnTransformer to handle the missing values for ea.col.
preprocessor = ColumnTransformer(
    transformers=[
        ('zero', zero_transformer, imp_zero_cols),
        ('mode', mode_transformer, imp_mode_cols),
        ('cats', categorical_transformer, ohe_cols)])

preprocessor

In [ ]:
## Fit and transform the data via the ColumnTransformer
X_train_tf = preprocessor.fit_transform(X_train)

In [ ]:
## Getting feature names from OHE
ohe_cat_names = preprocessor.named_transformers_['cats'].get_feature_names(ohe_cols)
ohe_cat_names

In [ ]:
## Generating list for column index
final_cols = [*imp_zero_cols, *imp_mode_cols, *ohe_cat_names]

In [ ]:
## Instantiating a DF storing the resulting transformed training data
X_train_tf_df = pd.DataFrame(X_train_tf, columns=final_cols, index=X_train.index)
X_train_tf_df

In [ ]:
## Transforming the test set
X_test_tf = preprocessor.transform(X_test)
X_test_tf

In [ ]:
## Creating DF for test data
X_test_tf_df = pd.DataFrame(X_test_tf, columns=final_cols, index=X_test.index)
X_test_tf_df

In [ ]:
sm_reg = sms.OLS(y_train, X_train_tf_df).fit()
sm_reg.summary()

In [ ]:
sm_reg.params

In [ ]:
## Creating full pipeline with ColumnTransformer and LinearRegression
reg_pipe = Pipeline(steps=[('preprocessor', preprocessor),
                      ('regressor', LinearRegression(fit_intercept=True))])

In [ ]:
## Fittng pipeline to the training data
reg_pipe.fit(X_train, y_train)

In [ ]:
## Running the pipeline on the training data
y_hat_train = reg_pipe.predict(X_train)
y_hat_train

In [ ]:
## Running the pipeline on the testing data
y_hat_test = reg_pipe.predict(X_test)
y_hat_test

In [ ]:
reg_pipe['regressor'].intercept_

In [ ]:
pd.set_option('display.float_format', lambda x: f'{x:,.2f}')

In [ ]:
pipe_coefs = pd.Series(reg_pipe['regressor'].coef_, index=final_cols)
pipe_coefs['intercept'] = reg_pipe['regressor'].intercept_
pipe_coefs.head(20)

In [ ]:
eval_perf_train(reg_pipe, X_train, y_train)

In [ ]:
eval_perf_test(reg_pipe, X_test, y_test)

In [ ]:
# ## Evaluating performance of training model

# train_mae = metrics.mean_absolute_error(y_train, y_hat_train)
# train_mse = metrics.mean_squared_error(y_train, y_hat_train)
# train_rmse = np.sqrt(metrics.mean_squared_error(y_train, y_hat_train))
# train_r = metrics.r2_score(y_train, y_hat_train)

# print('Evaluating performance of training data:\n')
# print(f'Mean Absolute Error: {train_mae:,.2f}')
# print(f'Mean Squared Error:  {train_mse:,.2f}\n')
# print(f'Root Mean Squared Error: {train_rmse:,.2f}')
# print(f'R-Square value for training data: {round(train_r,2)}')

In [ ]:
# ## Evaluating performance of testing model

# test_mae = metrics.mean_absolute_error(y_test, y_hat_test)
# test_mse = metrics.mean_squared_error(y_test, y_hat_test)
# test_rmse = np.sqrt(metrics.mean_squared_error(y_test, y_hat_test))
# test_r = metrics.r2_score(y_test, y_hat_test)

# print('Evaluating performance of testing data:\n')
# print(f'Mean Absolute Error: {test_mae:,.2f}')
# print(f'Mean Squared Error:  {test_mse:,.2f}\n')
# print(f'Root Mean Squared Error: {test_rmse:,.2f}')
# print(f'R-Square value for testing data: {round(test_r,2)}')

In [ ]:
# Creating first GridSearchCV to determine best parameters for LinReg

param_grid = {'fit_intercept': [True, False]}
ex_grid = GridSearchCV(LinearRegression(), param_grid, verbose=True,
                   scoring='r2')
ex_grid.fit(X_train_tf_df, y_train)
display(ex_grid.best_params_)
display(ex_grid.best_score_)
display(ex_grid.best_estimator_)
display(ex_grid.get_params())

In [ ]:
## Results in a DF format
# pd.DataFrame(reg.cv_results_).T

In [ ]:
## Performing GSCV on test data
y_hat_test = ex_grid.predict(X_test_tf_df)
y_hat_test

In [ ]:
eval_perf_test(ex_grid, X_test, y_test)

In [ ]:
# ## Evaluating model peformance

# test_mae = metrics.mean_absolute_error(y_test, y_hat_test)
# test_mse = metrics.mean_squared_error(y_test, y_hat_test)
# test_rmse = np.sqrt(metrics.mean_squared_error(y_test, y_hat_test))
# test_r = metrics.r2_score(y_test, y_hat_test)

# print('Evaluating performance of testing data.')
# print()
# print(f'Mean Absolute Error: {test_mae:,.2f}')
# print(f'Mean Squared Error:  {test_mse:,.2f}')
# print()
# print(f'Root Mean Squared Error: {test_rmse:,.2f}')
# print(f'R-Square value for testing data is {round(test_r,2)}.')

In [ ]:
## Running gridsearch on pipeline

param_grid = {'preprocessor__mode__imp_mode__strategy': ['mean','median',
                                                        'most_frequent']}
gs_imp_rmse = GridSearchCV(reg_pipe, param_grid, verbose=True,
                   scoring='neg_root_mean_squared_error')
gs_imp_rmse.fit(X_train, y_train)
display(gs_imp_rmse.best_score_, reg.best_params_)

In [ ]:
eval_perf_train(gs_imp_rmse, X_train, y_train)

In [ ]:
eval_perf_test(gs_imp_rmse, X_test, y_test)

In [ ]:
## Running gridsearch on pipeline

param_grid = {'preprocessor__mode__imp_mode__strategy': ['mean','median',
                                                        'most_frequent']}
gs_imp_r2 = GridSearchCV(reg_pipe, param_grid, verbose=True,
                   scoring='r2')
gs_imp_r2.fit(X_train, y_train)
display(gs_imp_r2.best_score_)
display(gs_imp_r2.best_params_)

In [ ]:
eval_perf_train(gs_imp_r2, X_train, y_train)

In [ ]:
eval_perf_test(gs_imp_r2, X_test, y_test)

In [ ]:
pd.DataFrame(gs_imp_r2.cv_results_)

# Old Code

In [ ]:
# ## Using OneHotEncoder to create columns for the categorical variables.

# ## Create list of categorical variables
# cat_feats = ['condition','grade', 'zipcode']

# ## Create the OHE without "drop='first" as it would throw an error in this case
# ohe = OneHotEncoder(handle_unknown = 'ignore', sparse=False)

# ## Using OHE on our categorical variables for training (NOT testing)
# train_ohe = ohe.fit_transform(X_train[cat_feats])
# test_ohe = ohe.transform(X_test[cat_feats])

In [ ]:
# ## Getting feature names from our list of categories
# feat_col_name = ohe.get_feature_names(cat_feats)

In [ ]:
# ## Creating dataframes from the results of our fit&transform and transform

# train_ohe_df = pd.DataFrame(train_ohe, columns=feat_col_name, index=X_train.index)

# test_ohe_df = pd.DataFrame(test_ohe, columns=feat_col_name, index=X_test.index)

In [ ]:
# ## Reviewing one of the resulting dataframes
# test_ohe_df

In [ ]:
# X_train

In [ ]:
# ## Combining the training dataframes after performing OHE
# X_train_ohe = pd.concat([X_train.drop(cat_feats, axis=1), train_ohe_df], axis=1)

In [ ]:
# ## Combining the testing dataframes after performing OHE
# X_test_ohe = pd.concat([X_test.drop(cat_feats, axis=1), test_ohe_df], axis=1)
# X_test_ohe

## Fitting the Model

In [ ]:
# ## Instantiating the model and fitting it
# lr = LinearRegression()

# lr.fit(X_train_ohe, y_train)

In [ ]:
# ## Checking the R^2 for train and test

# train_r = lr.score(X_train_ohe, y_train)
# print(f'R-Square value for training data is {round(train_r,3)}.')

# test_r = lr.score(X_test_ohe, y_test)
# print(f'R-Square value for test data is {round(test_r,3)}.')

# ## Getting model coefficients
# train_coef = pd.Series(lr.coef_, index=X_train_ohe.columns)
# train_coef['intercept'] = lr.intercept_

# ## Displaying resulting features and coefficients
# train_coef.sort_values(ascending=False)[:11]#.plot(kind='barh');

In [ ]:
# ## Determining predictions for training data

# y_train_pred = lr.predict(X_train_ohe)


# ## Determining predictions for test data

# y_test_pred = lr.predict(X_test_ohe)


## Evaluate Performance

In [ ]:
# ## Calculating scores for training model

# train_mae = metrics.mean_absolute_error(y_train, y_train_pred)
# train_mse = metrics.mean_squared_error(y_train, y_train_pred)
# train_rmse = np.sqrt(metrics.mean_squared_error(y_train, y_train_pred))

# print('Evaluating performance of training data.')
# print()
# print(f'Mean Absolute Error: {train_mae:,.2f}')
# print(f'Mean Squared Error:  {train_mse:,.2f}')
# print()
# print(f'Root Mean Squared Error: {train_rmse:,.2f}')
# train_r = lr.score(X_train_ohe, y_train)
# print(f'R-Square value for training data is {round(train_r,2)}.')

In [ ]:
# ## Calculating scores for test model

# test_mae = metrics.mean_absolute_error(y_test, y_test_pred)
# test_mse = metrics.mean_squared_error(y_test, y_test_pred)
# test_rmse = np.sqrt(metrics.mean_squared_error(y_test, y_test_pred))

# print(f'Mean Absolute Error: {test_mae:,.2f}')
# print(f'Mean Squared Error: {test_mse:,.2f}')
# print()
# print(f'Root Mean Squared Error: {test_rmse:,.2f}')
# test_r = lr.score(X_test_ohe, y_test)
# print(f'R-Square value for test data is {round(test_r,2)}.')

## Comparing Results

In [ ]:
# ## Calculating difference between RMSE(Training) and RMSE(Test)

# print('Our RMSE values for each set:')
# print(f'Train RMSE: {train_rmse:,.2f}')
# print(f'Test RMSE: {test_rmse:,.2f}')

# diff = train_rmse - test_rmse

# print()
# print(f'The difference in RMSE between our training and test sets is {diff:,.2f}.')

## Testing Assumptions

In [ ]:
## Determining residuals
residuals = (y_test - y_test_pred)
residuals

In [ ]:
## Plotting to test for normality
sns.histplot(data=residuals);

In [ ]:
## Checking the homoscedasticity of the new model
sns.residplot(x=y_test, y=residuals, lowess=True, color="g");

### Interpretation of Model Results

The model continues to do a job of explaining the variance between our selected features, with an r-squared value of .79/1.

The RMSE for both the training and test sets are below our target of 200,000, but higher than our model without the outliers. The difference between the two RMSEs did increase, indicating area for improvement.

The assumption tests confirm the assumptions that our residuals would be normally distributed and homoscedastic.

## Recursive Feature Elimination

The model using the original data performed worse in terms of RMSE, but maintained the R^2 value of .79.

Let's see if the RFECV has any impact this time.

In [ ]:
# ## Creating the settings for the selector
# selector = RFECV(estimator=lr, step=2, cv=3, scoring='neg_mean_squared_error', n_jobs=-1, verbose=1)

In [ ]:
# ## Running it on our training data
# selector.fit(X_train_ohe, y_train)

In [ ]:
# ## Dividing the data into two separate variables for inspection
# selected_columns = X_train_ohe.columns[selector.get_support()]
# removed_columns = X_train_ohe.columns[~selector.get_support()]

In [ ]:
# ## Inspecting the columns with the greatest impact according to the RSECV method
# list(selected_columns.values)

In [ ]:
# ## Testing the new list of columns

# ## Instantiating a new regression model
# lr_rfe = LinearRegression()

# ## Fitting the model to the new data
# lr_rfe.fit(X_train_ohe[selected_columns], y_train)

# # Using the model to create new predictions
# y_rfe = lr_rfe.predict(X_train_ohe[selected_columns])

# # Determining RMSE for new list of columns and TRAIN data
# trainRFE_rmse = np.sqrt(metrics.mean_squared_error(y_train, y_rfe))

# print(f'Training Root Mean Squared Error: {trainRFE_rmse:,.2f}')

# # Determining RMSE for new list of columns and TEST data
# y_pred_rfe = lr_rfe.predict(X_test_ohe[selected_columns])

# testRFE_rmse = np.sqrt(metrics.mean_squared_error(y_test, y_pred_rfe))

# print(f'Testing Root Mean Squared Error: {testRFE_rmse:,.2f}')

# diff2 = trainRFE_rmse - testRFE_rmse
# print()
# print(f'The difference in RMSE between our training and test sets is {diff2:,.2f}.')

In [ ]:
# test_rsq = lr_rfe.score(X_test_ohe[selected_columns], y_test)
# print(f'R-Square value for testing data after RFECV is {round(test_rsq,2)}.')

---
**Comparing RMSEs from original to RFECV:**

* **Pre-RFECV model results:**
   * The R^2 value for both training and testing data was .79 (significant as it is over the .70 threshold).
   * *Our RMSE values for each set:*
        * Train RMSE: 117,679.79
        * Test RMSE: 122,263.56
    * The difference in RMSE between our training and test sets is **-4,583.77**.


 * **Post-RFECV model results:**
     * The R^2 value for the test data is .48 (.31 lower than other model, and far below threshold).
     * *Our RMSE values for each set:*
         * Training Root Mean Squared Error: 182,158.41
         * Testing Root Mean Squared Error: 191,360.88
     * The difference in RMSE between our training and test sets is **-9,202.47**.

---

**Final Analysis of RFE method:** The RFE results produced a large decrease in model perfomance as determined by the lower R^2 and lower RMSE values. I will continue with the first model for future predictions.

---

# **Pickling the Model for Reuse**

In [ ]:
# ## Setting coefficients and target
# X = df[['was_renovated','waterfront', 'zipcode',
#             'yrs_since_reno', 'bedrooms', 'bathrooms', 'condition','grade']].copy()
# y = df['price'].copy()

In [ ]:
# ## Create list of categorical variables
# cat_feats = ['condition','grade', 'zipcode']

# ## Create the OHE without "drop='first" as it would throw an error in this case
# ohe = OneHotEncoder(handle_unknown = 'ignore', sparse=False)

# ## Using OHE on our categorical variables for training (NOT testing)
# final_ohe = ohe.fit_transform(X[cat_feats])

In [ ]:
# ## Getting feature names from our list of categories
# feat_col_name = ohe.get_feature_names(cat_feats)

In [ ]:
# ## Creating DF of final results
# penultimate_ohe_df = pd.DataFrame(final_ohe, columns=feat_col_name, index=X.index)

In [ ]:
# ## Updating final DF
# final_ohe_df = pd.concat([X.drop(cat_feats, axis=1), penultimate_ohe_df], axis=1)

In [ ]:
# ## Instantiating and fitting the regression
# lr_final = LinearRegression()

# lr_final = lr_final.fit(final_ohe_df, y)

In [ ]:
# # Generating predictions
# final_pred = lr_final.predict(final_ohe_df)

In [ ]:
# ## Confirming predictions work
# display(final_pred)

In [ ]:
# # Saving model to .pickle file
# import pickle

# pickle_out = open("final_model.pickle","wb")
# pickle.dump(lr_final, pickle_out)
# pickle_out.close()